File: 2021/go/day2/part2/part2.go
package part2

import (
	"bufio"
	"strconv"
	"strings"
)

type Command struct {
	Operation string
	Value     int
}

func Solve(input string) string {
	scanner := bufio.NewScanner(strings.NewReader(input))
	var commands []*Command
	for scanner.Scan() {
		s := strings.Split(scanner.Text(), " ")
		v, _ := strconv.Atoi(s[1])
		commands = append(commands, &Command{
			Operation: s[0],
			Value:     v,
		})
	}
	return strconv.Itoa(dive(commands))
}

func dive(commands []*Command) int {
	horizontal, depth, aim := 0, 0, 0
	for _, c := range commands {
		switch c.Operation {
		case "forward":
			horizontal += c.Value
			depth += c.Value * aim
		case "up":
			aim -= c.Value
		case "down":
			aim += c.Value
		}
	}
	return horizontal * depth
}

File: 2021/elixir/apps/day2/lib/part2.ex
defmodule Day2Part2 do
  @moduledoc """
  Day2 - Part2
  """

  defmodule SubmarineState do
    defstruct horizontal: 0, depth: 0, aim: 0
  end

  def solve(input) do
    input
    |> String.split("\n", trim: true)
    |> Enum.reduce(%SubmarineState{}, &dive/2)
    |> then(fn %SubmarineState{horizontal: h, depth: d} -> h * d end)
  end

  defp dive("forward " <> value, %SubmarineState{horizontal: h, depth: d, aim: a}) do
    v = String.to_integer(value)
    %SubmarineState{horizontal: h + v, depth: d + v * a, aim: a}
  end

  defp dive("up " <> value, %SubmarineState{horizontal: h, depth: d, aim: a}) do
    %SubmarineState{horizontal: h, depth: d, aim: a - String.to_integer(value)}
  end

  defp dive("down " <> value, %SubmarineState{horizontal: h, depth: d, aim: a}) do
    %SubmarineState{horizontal: h, depth: d, aim: a + String.to_integer(value)}
  end
end
